File: backend/src/main/java/com/whiteblue/backend/security/OAuth/OAuthUserService.java
package com.whiteblue.backend.security.OAuth;

import com.whiteblue.backend.domain.user.User;
import com.whiteblue.backend.domain.user.UserRepository;
import com.whiteblue.backend.domain.user.UserRole;
import lombok.RequiredArgsConstructor;
import org.springframework.security.oauth2.client.userinfo.DefaultOAuth2UserService;
import org.springframework.security.oauth2.client.userinfo.OAuth2UserRequest;
import org.springframework.security.oauth2.core.OAuth2AuthenticationException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Map;
import java.util.Optional;

@RequiredArgsConstructor
@Transactional
@Service
public class OAuthUserService extends DefaultOAuth2UserService {
    private final UserRepository userRepository;

    public OAuthUser loadUser(OAuth2UserRequest userRequest) throws OAuth2AuthenticationException {
        Map<String, Object> attributes = super.loadUser(userRequest)
                                              .getAttributes();
        Map<String, Object> kakao_account = (Map<String, Object>) attributes.get("kakao_account");
        Map<String, Object> profile = (Map<String, Object>) kakao_account.get("profile");
        String username = (String) kakao_account.get("email");
        String name = (String) profile.get("nickname");
        String image = (Boolean) profile.get("is_default_image") ? null : (String) profile.get("profile_image_url");

        Optional<User> userOptional = userRepository.findByUsername(username);
        User user;

        if (userOptional.isPresent()) {
            user = userOptional.get();
            user.setName(name);
            user.setImage(image);
        } else {
            user = userRepository.save(User.builder()
                                           .username(username)
                                           .name(name)
                                           .role(UserRole.USER)
                                           .image(image)
                                           .build());
        }

        return new OAuthUser(user, attributes);
    }

}

File: backend/src/main/kotlin/com/whiteblue/backend/security/oAuth/OAuthUserService.kt
package com.whiteblue.backend.security.oAuth

import com.whiteblue.backend.domain.user.User
import com.whiteblue.backend.domain.user.UserRepository
import com.whiteblue.backend.domain.user.UserRole
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.oauth2.client.userinfo.DefaultOAuth2UserService
import org.springframework.security.oauth2.client.userinfo.OAuth2UserRequest
import org.springframework.security.oauth2.core.user.OAuth2User
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.*

@Transactional
@Service
class OAuthUserService(private val userRepository: UserRepository):
    DefaultOAuth2UserService() {
    override fun loadUser(userRequest: OAuth2UserRequest): OAuth2User {
        val attributes: Map<String, Any> = super.loadUser(userRequest).attributes
        val account = attributes["kakao_account"] as Map<*, *>
        val profile = account["profile"] as Map<*, *>
        val username = account["email"] as String
        val name = profile["nickname"] as String
        val image = if (profile["is_default_image"] as Boolean) null else profile["profile_image_url"] as String

        val user = userRepository.findByUsername(username)
            ?.let {
                it.name = name
                it.image = image
                userRepository.save(it)
            } ?: run { userRepository.save(User(username = username, name = name, image = image, role = UserRole.USER)) }

        return OAuthUser(
            id = user.id!!,
            username = user.username,
            name = user.name,
            authorities = Collections.singletonList(
                SimpleGrantedAuthority(
                    user.role.roleName
                )
            ),
            attributes = attributes
        )
    }
}
